# 03 - Data Manipulation

## Import Package

In [1]:
# Import package as pandas does not ship along as native package
# use pip install pandas if not installed yet, or follow setup guide

import pandas as pd
import matplotlib.pyplot as plt

## A Simple Example: How many Pokemon are there for each type?

In [2]:
dataframe = pd.read_csv("https://raw.githubusercontent.com/KianYang-Lee/pandas-tutorial/main/datasets/Pokemon.csv")
dataframe.head()

#                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False

## How DataFrame is Indexed

In [3]:
# Every DataFrame at least has one row index and
# one column index
dataframe.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

In [4]:
dataframe.index

RangeIndex(start=0, stop=800, step=1)

## Accessing Column(s)

In [5]:
# Accessing column(s)
# dot notation to access attribute of object
dataframe.Name

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

In [6]:
# similar operation but using indexing operator - square bracket
dataframe["Name"]

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

In [7]:
# select multiple columns
dataframe[["Name", "HP", "Attack", "Defense"]]

Name  HP  Attack  Defense
0                Bulbasaur  45      49       49
1                  Ivysaur  60      62       63
2                 Venusaur  80      82       83
3    VenusaurMega Venusaur  80     100      123
4               Charmander  39      52       43
..                     ...  ..     ...      ...
795                Diancie  50     100      150
796    DiancieMega Diancie  50     160      110
797    HoopaHoopa Confined  80     110       60
798     HoopaHoopa Unbound  80     160       60
799              Volcanion  80     110      120

[800 rows x 4 columns]

In [8]:
# drilling down
dataframe["Name"][0]

'Bulbasaur'

## Indexing

In [9]:
# index-based selection - based on index position
# select second row
dataframe.iloc[1, :]

#                   2
Name          Ivysaur
Type 1          Grass
Type 2         Poison
Total             405
HP                 60
Attack             62
Defense            63
Sp. Atk            80
Sp. Def            80
Speed              60
Generation          1
Legendary       False
Name: 1, dtype: object

In [10]:
# select last column
dataframe.iloc[:, -1]

0      False
1      False
2      False
3      False
4      False
       ...  
795     True
796     True
797     True
798     True
799     True
Name: Legendary, Length: 800, dtype: bool

In [11]:
# label-based selection - based on index value
# select Name column
dataframe.loc[:, "Name"]

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

In [12]:
# select second row
dataframe.loc[1, :]

#                   2
Name          Ivysaur
Type 1          Grass
Type 2         Poison
Total             405
HP                 60
Attack             62
Defense            63
Sp. Atk            80
Sp. Def            80
Speed              60
Generation          1
Legendary       False
Name: 1, dtype: object

In [13]:
# select multiple columns
dataframe.loc[:, ["Name", "HP", "Attack", "Defense"]]

Name  HP  Attack  Defense
0                Bulbasaur  45      49       49
1                  Ivysaur  60      62       63
2                 Venusaur  80      82       83
3    VenusaurMega Venusaur  80     100      123
4               Charmander  39      52       43
..                     ...  ..     ...      ...
795                Diancie  50     100      150
796    DiancieMega Diancie  50     160      110
797    HoopaHoopa Confined  80     110       60
798     HoopaHoopa Unbound  80     160       60
799              Volcanion  80     110      120

[800 rows x 4 columns]

In [14]:
# region Index can be changes, like everything else in life
df_with_name_index = dataframe.set_index("Name", drop=True)
df_with_name_index.loc["Charizard"]

#                  6
Type 1          Fire
Type 2        Flying
Total            534
HP                78
Attack            84
Defense           78
Sp. Atk          109
Sp. Def           85
Speed            100
Generation         1
Legendary      False
Name: Charizard, dtype: object

## Conditional Selection

In [15]:
# Conditional selection that fulfills certain criteria
# setting a condition to return a Boolean Series
min_HP_condition = dataframe.HP > 200

In [16]:
# this line will not work, do you know why? [HINTS: recall nature of iLoc]
# dataframe.iloc[min_HP_condition]

In [17]:
# conditional selection of rows that satisfies the condition
dataframe.loc[min_HP_condition]

#     Name  Type 1 Type 2  Total   HP  Attack  Defense  Sp. Atk  \
121  113  Chansey  Normal    NaN    450  250       5        5       35   
261  242  Blissey  Normal    NaN    540  255      10       10       75   

     Sp. Def  Speed  Generation  Legendary  
121      105     50           1      False  
261      135     55           2      False

In [18]:
# conditional selection on multiple criteria
min_attack_condition = dataframe.Attack > 100
min_defense_condition = dataframe.Defense > 100
dataframe.loc[min_attack_condition & min_defense_condition]

#                       Name   Type 1    Type 2  Total   HP  Attack  \
7      6  CharizardMega Charizard X     Fire    Dragon    634   78     130   
12     9    BlastoiseMega Blastoise    Water       NaN    630   79     103   
82    76                      Golem     Rock    Ground    495   80     120   
107   99                    Kingler    Water       NaN    475   55     130   
120  112                     Rhydon   Ground      Rock    485  105     130   
137  127          PinsirMega Pinsir      Bug    Flying    600   65     155   
141  130      GyaradosMega Gyarados    Water      Dark    640   95     155   
152  141                   Kabutops     Rock     Water    495   60     115   
224  208        SteelixMega Steelix    Steel    Ground    610   75     125   
229  212          ScizorMega Scizor      Bug     Steel    600   70     150   
232  214    HeracrossMega Heracross      Bug  Fighting    600   80     185   
251  232                    Donphan   Ground       NaN    500   90     120   
267  248                  Tyranitar     Rock      Dark    600  100     134   
268  248    TyranitarMega Tyranitar     Rock      Dark    700  100     164   
283  260      SwampertMega Swampert    Water    Ground    635  100     150   
329  303          MawileMega Mawile    Steel     Fairy    480   50     105   
332  306                     Aggron    Steel      Rock    530   70     110   
333  306          AggronMega Aggron    Steel       NaN    630   70     140   
366  334        AltariaMega Altaria   Dragon     Fairy    590   75     110   
402  367                    Huntail    Water       NaN    485   55     104   
409  373    SalamenceMega Salamence   Dragon    Flying    700   95     145   
412  376                  Metagross    Steel   Psychic    600   80     135   
413  376    MetagrossMega Metagross    Steel   Psychic    700   80     145   
423  383                    Groudon   Ground       NaN    670  100     150   
424  383      GroudonPrimal Groudon   Ground      Fire    770  100     180   
434  389                   Torterra    Grass    Ground    525   95     109   
494  445      GarchompMega Garchomp   Dragon    Ground    700  108     170   
500  450                  Hippowdon   Ground       NaN    525  108     112   
511  460    AbomasnowMega Abomasnow    Grass       Ice    594   90     132   
515  464                  Rhyperior   Ground      Rock    535  115     140   
521  470                    Leafeon    Grass       NaN    525   65     110   
538  481                    Mesprit  Psychic       NaN    580   80     105   
540  483                     Dialga    Steel    Dragon    680  100     120   
543  486                  Regigigas   Normal       NaN    670  110     160   
552  493                     Arceus   Normal       NaN    720  120     120   
584  525                    Boldore     Rock       NaN    390   70     105   
585  526                   Gigalith     Rock       NaN    515   85     135   
626  565                 Carracosta    Water      Rock    495   74     108   
650  589                 Escavalier      Bug     Steel    495   70     135   
693  632                     Durant      Bug     Steel    484   58     109   
707  644                     Zekrom   Dragon  Electric    680  100     150   
720  652                 Chesnaught    Grass  Fighting    530   88     107   
749  680                   Doublade    Steel     Ghost    448   59     110   
759  689                 Barbaracle     Rock     Water    500   72     105   
767  697                  Tyrantrum     Rock    Dragon    521   82     121   
789  713                    Avalugg      Ice       NaN    514   95     117   
796  719        DiancieMega Diancie     Rock     Fairy    700   50     160   
799  721                  Volcanion     Fire     Water    600   80     110   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
7        111      130       85    100           1      False  
12       120      135      115     78           1      Fal

In [19]:
dataframe.loc[min_attack_condition | min_defense_condition]

#                       Name   Type 1  Type 2  Total  HP  Attack  \
3      3      VenusaurMega Venusaur    Grass  Poison    625  80     100   
7      6  CharizardMega Charizard X     Fire  Dragon    634  78     130   
8      6  CharizardMega Charizard Y     Fire  Flying    634  78     104   
12     9    BlastoiseMega Blastoise    Water     NaN    630  79     103   
19    15      BeedrillMega Beedrill      Bug  Poison    495  65     150   
..   ...                        ...      ...     ...    ...  ..     ...   
795  719                    Diancie     Rock   Fairy    600  50     100   
796  719        DiancieMega Diancie     Rock   Fairy    700  50     160   
797  720        HoopaHoopa Confined  Psychic   Ghost    600  80     110   
798  720         HoopaHoopa Unbound  Psychic    Dark    680  80     160   
799  721                  Volcanion     Fire   Water    600  80     110   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
3        123      122      120     80           1      False  
7        111      130       85    100           1      False  
8         78      159      115    100           1      False  
12       120      135      115     78           1      False  
19        40       15       80    145           1      False  
..       ...      ...      ...    ...         ...        ...  
795      150      100      150     50           6       True  
796      110      160      110    110           6       True  
797       60      150      130     70           6       True  
798       60      170      130     80           6       True  
799      120      130       90     70           6       True  

[245 rows x 13 columns]

## Changing Value(s)

In [20]:
# Changing a value
# able to access specific value in the DataFrame means that
# you can also change the value
dataframe.loc[0, "Name"] = "Rilakuma"

In [21]:
# Changing multiple values at once
# mapping from one value to another for a Series
dataframe.HP.map(lambda x: 0)

0      0
1      0
2      0
3      0
4      0
      ..
795    0
796    0
797    0
798    0
799    0
Name: HP, Length: 800, dtype: int64

In [22]:
# applying a transformation on a DataFrame
# transform each column by passing "columns" or 1 argument to axis parameter
dataframe.apply(lambda x: x * 2, axis="columns")

#                                        Name          Type 1  \
0       2                            RilakumaRilakuma      GrassGrass   
1       4                              IvysaurIvysaur      GrassGrass   
2       6                            VenusaurVenusaur      GrassGrass   
3       6  VenusaurMega VenusaurVenusaurMega Venusaur      GrassGrass   
4       8                        CharmanderCharmander        FireFire   
..    ...                                         ...             ...   
795  1438                              DiancieDiancie        RockRock   
796  1438      DiancieMega DiancieDiancieMega Diancie        RockRock   
797  1440      HoopaHoopa ConfinedHoopaHoopa Confined  PsychicPsychic   
798  1440        HoopaHoopa UnboundHoopaHoopa Unbound  PsychicPsychic   
799  1442                          VolcanionVolcanion        FireFire   

           Type 2  Total   HP  Attack  Defense  Sp. Atk  Sp. Def  Speed  \
0    PoisonPoison    636   90      98       98      130      130     90   
1    PoisonPoison    810  120     124      126      160      160    120   
2    PoisonPoison   1050  160     164      166      200      200    160   
3    PoisonPoison   1250  160     200      246      244      240    160   
4             NaN    618   78     104       86      120      100    130   
..            ...    ...  ...     ...      ...      ...      ...    ...   
795    FairyFairy   1200  100     200      300      200      300    100   
796    FairyFairy   1400  100     320      220      320      220    220   
797    GhostGhost   1200  160     220      120      300      260    140   
798      DarkDark   1360  160     320      120      340      260    160   
799    WaterWater   1200  160     220      240      260      180    140   

     Generation  Legendary  
0             2          0  
1             2          0  
2             2          0  
3             2          0  
4             2          0  
..          ...        ...  
795          12          2  
796          12          2  
797          12          2  
798          12          2  
799          12          2  

[800 rows x 13 columns]

## Practice Section

1. Subset the following columns and assign them to another variable.
- Sp. Atk
- Sp. Def
- Speed

2. Find all attributes in Question 1 of the following Pokemon and assign to a variable respectively.
eg. Diglet 35 10 70
- Mewtwo
- Ratata
- Pikachu

3. Assign the information listed in Question 1 of the Pokemon with the following index as `DataFrame` to another variable.
- 50
- 100
- 150

4. Modify the `Attack` and `Defense` of `Snorlax` to 300.
5. Create a subset of the original `DataFrame` containing only the First Generation Pokemon, and assign it to another variable.
6. Create a subset of the original `DataFrame` containing only the Second Generation Pokemon with `Sp. Atk` **and** `Sp. Def` of at least 100. Assign the `DataFrame` to another variable.
7. Create a subset of the original `DataFrame` containing only the Second Generation Pokemon with `Sp. Atk` **or** `Sp. Def` of at least 100. Assign the `DataFrame` to another variable.

## Useful Resources
A great discussion on the difference between `map`, `applymap` and `apply` methods can be found on this [post](https://stackoverflow.com/questions/19798153/difference-between-map-applymap-and-apply-methods-in-pandas) in Stackoverflow.

**Copyright (C) 2021  Lee Kian Yang**

This program is licensed under MIT license.